snap7/client_async.py

```diff
@@ -173,3 +173,59 @@ async def as_download(self, data, block_num=-1, timeout=1):
         if request_in_time is False:
             return None
         return data
+
+    async def as_read_area(self, area, dbnumber, start, size, timeout=1):
+        """This is the main async function to read data from a PLC, made asycnio compatible.
+        With it you can read DB, Inputs, Outputs, Merkers, Timers and Counters.
+
+        :param timeout: max time waiting for response, 1 Second default
+        :param area: chosen memory_area
+        :param dbnumber: The DB number, only used when area= S7AreaDB
+        :param start: offset to start writing
+        :param size: number of units to read
+        """
+        assert area in snap7.snap7types.areas.values()
+        if area == snap7.snap7types.S7AreaTM:
+            wordlen = snap7.snap7types.S7WLTimer
+        elif area == snap7.snap7types.S7AreaCT:
+            wordlen = snap7.snap7types.S7WLCounter
+        else:
+            wordlen = snap7.snap7types.S7WLByte
+        type_ = snap7.snap7types.wordlen_to_ctypes[wordlen]
```

Comment:
i think you dont need to replicate the code from client.py but just call it and make a python async wrapper around it.

Reply:
I'm not sure if a wrapper is the solution, because check_error() would run twice with perhaps two different solutions.
In the sync version, the result is never awaited (just hoped that a legit result received in the time between the lines), so I'm not sure what may could happened there.
But you are still right, it's still not fitting in the DRY (don't repeat yourself) -concept, I will think about a better solution about that.

Reply:
why would check_error run twice?

Reply:
Afaik if you wrap a function, the wrapper can perform prolog and epilog stuff only. So if I would wrap `as_area_read()` from `client.py` I would dive through the wrapper, to the function, execute it all way, do the epilog stuff (which could be the asyncio stuff) and exit it the whole statementblock again.
But `as_area_read()` already contain `check_error(result, context="client")` and adding       
```
 request_in_time = await self.as_check_and_wait(timeout)
 if request_in_time is False:
     return None
 check_error(result, context="client")
 return bytearray(data)
```
in the epilog would be `check_error()` again.

Or do you tend to something else?

Reply:
in the end, you just do a call to the library that you want to have error-checked right? You can use the existing library structure for that. Your async code is a wrapper around that.

Reply:
I see still a difference (and not good wrappable solution) between the sync:
```
        result = self._library.Cli_AsReadArea(self._pointer, area, dbnumber, start, size, wordlen, byref(data))
        check_error(result, context="client")
        return bytearray(data)
```
and my async version:
```
       result = self._library.Cli_AsReadArea(self._pointer, area, dbnumber, start, size, wordlen, byref(data))
        request_in_time = await self.as_check_and_wait(timeout)
        if request_in_time is False:
            return None
        check_error(result, context="client")
        return bytearray(data)
```
In the first example you are sending the read command, and then execute check_error directly afterwards. No check if AsReadArea was received.

In my example you are sending the request, wait until a timeout/execution and then check_error.

This leads to two different `results` (one checked/legit result, one not checked/faulty/perhaps right result)  and so `check_error()` may return different results, wouldn't it? 
I want explicitly wait until `CheckAsCompletion()` shows green light for a valid result, before I want to execute check_error(), because your result and my result will mostly be different.

If I would just make a decorator at `as_area_read()`,  the first `check_error()` may throw a Snap7Exception, even if I wait until a legit result afterwards in my wrapper code and the second `check_error()` there would be fine with that.

I will commit an idea the next hours, perhaps we will agree on that. :smile:  
Sorry for all the trouble I cause here.

Reply:
you should always do the check_error on the result of the call, since the library will return an error code anyway. This will not change. Do not await in between.

Reply:
Sorry for that stupid question: So just the call itself is checked with `check_error()`, not the received data from the PLC? Because if this is the case, I would finally understand your view and then, of course, a wrapper could be the most elegant solution.

Reply:
yes, the check_error() parses an error code returned by the call. You should always check if the low level call is successful.